include/seqan3/search/kmer_index/shape_iterator.hpp

```diff
@@ -0,0 +1,473 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
+ * \author Vinzenz May <vinzenz.may AT fu-berlin.de>
+ * \brief Provides the seqan3::shape_iterator.
+ */
+
+#pragma once
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/search/kmer_index/shape.hpp>
+#include <seqan3/std/ranges>
+
+namespace seqan3
+{
+
+/*!\brief Iterator for calculating hash values via a given seqan3::shape.
+ * \tparam it_t Type of iterator on text. Must model std::ForwardIterator. Value type must model seqan3::Semialphabet.
+*/
+template <std::ForwardIterator it_t>
+    //!\cond
+    requires Semialphabet<value_type_t<it_t>>
+    //!\endcond
+class shape_iterator
+{
+public:
+    //!\brief Type for distances between iterators.
+    using difference_type = typename it_t::difference_type;
+    //!\brief Value type of this iterator.
+    using value_type = size_t;
+    //!\brief The reference type of this iterator.
+    using reference = shape_iterator &;
+    //!\brief The const reference type of this iterator.
+    using const_reference = shape_iterator const &;
+    //!\brief Tag this class depending on which concept it_t models.
+    using iterator_category = std::conditional_t<std::RandomAccessIterator<it_t>,
+                                                 typename std::random_access_iterator_tag,
+                                                 std::conditional_t<std::BidirectionalIterator<it_t>,
+                                                                    typename std::bidirectional_iterator_tag,
+                                                                    typename std::forward_iterator_tag>>;
+
+private:
+    //!\brief The size type of this iterator.
+    using position_type = std::make_unsigned_t<difference_type>;
+    //!\brief The alphabet type of the passed iterator.
+    using alphabet_t = value_type_t<it_t>;
+
+    //!\brief The hash value.
+    size_t hash_value{0};
+
+    //!\brief The seqan3::shape to use for hashing.
+    shape s;
```

Comment:
I would prefer longer variable names. grep and replace is tedious otherwise and it is not expressive.. maybe `shape_` or `kmer_shape`?